### PR TITLE
(Do not merge) Partially implement blocking as mixed content

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/16320)

### CODE REVIEW

**Comment by nox** [comment]
This covers the script and style cases from https://github.com/servo/servo/issues/8582, AFAICT.

**Comment by jdm** [comment]
This should modify script/layout_image.rs as well.

**Comment by jdm** [comment]
I'm interested in having @avadacatavra review this.

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -1102,6 +1105,18 @@ fn http_network_fetch(request: &Request,
     response.headers = res.response.headers.clone();
     response.referrer = request.referrer.to_url().cloned();
 
+    {
+        let http_message = res.response.get_ref().downcast_ref::<Http11Message>().unwrap();
```
Comment (avadacatavra) [inline components/net/http_loader.rs]: all of this is to figure out if the response is served over https, right?

Comment (nox) [inline components/net/http_loader.rs]: Yes. I could more simply just check that the request's URL scheme was "https", but this checks that the very stream that handles the bits over the wire actually uses an `HttpsStream`, so I think that is better. And we will need that code when we will finally look at the actual cypher anyway.

Comment (avadacatavra) [inline components/net/http_loader.rs]: i agree :)

File: components/net/Cargo.toml [hunk 2]
```diff
@@ -15,9 +15,9 @@ brotli = "1.0.6"
 cookie = "0.6"
 devtools_traits = {path = "../devtools_traits"}
 flate2 = "0.2.0"
-hyper = "0.10"
+hyper = "0.10.8"
```
Comment (avadacatavra) [inline components/net/Cargo.toml]: why do we need to specify 0.10.8 instead of just 0.10?

File: components/net/Cargo.toml [hunk 3]
```diff
@@ -15,9 +15,9 @@ brotli = "1.0.6"
...
 hyper_serde = "0.6"
-hyper-openssl = "0.2.2"
+hyper-openssl = "0.2.6"
```
Comment (avadacatavra) [inline components/net/Cargo.toml]: same question--why do we need to specify past 0.2?

File: components/net/fetch/methods.rs [hunk 4]
```diff
@@ -255,7 +271,12 @@ pub fn main_fetch(request: &mut Request,
         // TODO: handle blocking by content security policy.
         let blocked_error_response;
         let internal_response =
-            if should_replace_with_nosniff_error {
+            if should_replace_with_mixed_content_error {
```
Comment (jdm) [inline components/net/fetch/methods.rs]: Please remove the TODO comment about mixed content above this line.

File: components/net/Cargo.toml (hunk 2 above)
Comment (nox) [inline components/net/Cargo.toml]: Good practice, the `get_ref` functions are in hyper 0.10.8.

File: components/net/Cargo.toml (hunk 3 above)
Comment (nox) [inline components/net/Cargo.toml]: Same reason, though I ended up not using the `lock` function I introduced in `0.2.6`, given we don't need to look at the cipher yet.

File: components/net/fetch/methods.rs [hunk 5]
```diff
@@ -651,3 +672,171 @@ fn is_bad_port(port: u16) -> bool {
...
+    // Step 3.
+    false
+}
+
+/// https://w3c.github.io/webappsec-mixed-content/#a-priori-authenticated-url
+fn is_a_priori_authenticated_url(url: &ServoUrl) -> bool {
+    is_potentially_trustworthy_url(url.as_url()) || url.scheme() == "data"
+}
+
+/// https://www.w3.org/TR/secure-contexts/#is-url-trustworthy
```
Comment (jdm) [inline components/net/fetch/methods.rs]: https://w3c.github.io/webappsec-secure-contexts/#is-url-trustworthy

File: components/net/fetch/methods.rs [hunk 6]
```diff
@@ -651,3 +672,171 @@ fn is_bad_port(port: u16) -> bool {
...
+    // Step 2.
+    if url.as_str() == "about:blank" || url.as_str() == "about:srcdoc" {
+        return true;
+    }
+
+    // Step 3.
+    is_potentially_trustworthy_origin(url.origin())
+}
+
+/// https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy
```
Comment (jdm) [inline components/net/fetch/methods.rs]: https://w3c.github.io/webappsec-secure-contexts/#is-origin-trustworthy

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/xmlhttprequest.rs [hunk 7]
```diff
@@ -583,9 +583,11 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
             method: self.request_method.borrow().clone(),
             url: self.request_url.borrow().clone().unwrap(),
             headers: (*self.request_headers.borrow()).clone(),
+            client: Some(self.global().get_request_client()),
             unsafe_request: true,
             // XXXManishearth figure out how to avoid this clone
             body: extracted_or_serialized.as_ref().map(|e| e.0.clone()),
+            // TODO: handle request's client.
```
Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: This is handled now.

File: components/net/fetch/methods.rs [hunk 8]
```diff
@@ -651,3 +671,154 @@ fn is_bad_port(port: u16) -> bool {
...
+    // Step 2.1.
+    // TODO: handle user agent configured to block optionally-blockable
+    // mixed content.
+
+    // Step 2.2.
+    // TODO: handle request’s client’s strict mixed content checking flag.
+
+    // Steps 3-4.
+    // TODO: handle whether response is an opaque filtered response.
+    // https://github.com/servo/servo/issues/16355
```
Comment (jdm) [inline components/net/fetch/methods.rs]: This hasn't been addressed yet.

File: components/script/dom/window.rs [hunk 9]
```diff
@@ -386,6 +386,27 @@ impl Window {
...
+                    responsible_document = ScriptThread::find_document(parent_id).unwrap();
+                },
+                _ => return false,
+            }
+        }
+    }
+
+    /// https://w3c.github.io/webappsec-mixed-content/#categorize-settings-object
+    pub fn target_browsing_context_has_parent_browsing_context(&self) -> bool {
+        self.is_top_level()
```
Comment (jdm) [inline components/script/dom/window.rs]: This check is backwards.

File: components/script/dom/window.rs [hunk 10]
```diff
@@ -386,6 +386,27 @@ impl Window {
...
     }
+
+    /// https://w3c.github.io/webappsec-mixed-content/#categorize-settings-object
+    pub fn prohibit_mixed_security_contexts(&self) -> bool {
+        let mut responsible_document = self.Document();
+        loop {
+            if responsible_document.https_state() != HttpsState::None {
+                return true;
+            }
+            match responsible_document.window().parent_info() {
```
Comment (jdm) [inline components/script/dom/window.rs]: This is broken for cross-origin iframes. We will need to talk to the constellation to make this work correctly.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by avadacatavra** [comment]
@nox what's the status on this?

**Comment by jdm** [comment]
Closing due to inactivity; tracking in https://github.com/servo/servo/issues/19243.

**Pull request closed without merging**